src/librustc_driver/driver.rs

```diff
@@ -1275,9 +1275,9 @@ pub fn build_output_filenames(input: &Input,
                                            .values()
                                            .filter(|a| a.is_none())
                                            .count();
-            let ofile = if unnamed_output_types > 1 {
-                sess.warn("ignoring specified output filename because multiple outputs were \
-                           requested");
+            let ofile = if unnamed_output_types > 1 && sess.opts.output_types.get(&OutputType::Exe).is_none() {
```

Comment:
Instead of `.get(...).is_none()` I would do `!sess.opts.output_types.contains_key(&OutputType::Exe)` (clearer as to the meaning).